### ISSUE TITLE
Create RHI vertex and index buffers asynchronously

### ISSUE THREAD

**Issue by nithinp7** [CesiumGS/cesium-unreal#982]
In #975, we implemented asynchronous RHI texture creation after discovering the [RHIAsyncCreateTexture2D](https://docs.unrealengine.com/4.26/en-US/API/Runtime/RHI/RHIAsyncCreateTexture2D/) function in Unreal. Similarly, there exist [RHIAsyncCreateIndexBuffer](https://docs.unrealengine.com/4.26/en-US/API/Runtime/RHI/RHIAsyncCreateIndexBuffer/) and [RHIAsyncCreateVertexBuffer](https://docs.unrealengine.com/4.26/en-US/API/Runtime/RHI/RHIAsyncCreateVertexBuffer/) that seem to suggest we can do a similar optimization for index and vertex buffers. We may in the process even find ways to directly upload quantized buffers. With run-of-the mill tilesets, this optimization may not improve much - but high-quality photogrammetry tilesets (e.g., centimeter-precise models) will almost certainly bottleneck during render thread vertex buffer creation.

As with the RHI-level texture creation PR, we will have to determine how to "link" the Unreal user-level static mesh object to the RHI-level resource. There will probably be a bunch of edge cases and varying support across platforms, but generally I think this should be possible.

**Comment by RaiaN** [comment]
I thought my comment might be useful for your future work. Previously I dealt with partial Vertex & Index buffer uploads in UE4. Apparently UE RHI layer didn't really use offset for `static` buffers so I ended up modifying that part of RHI (D3D11 only) i.e. source code. Hopefully no more changes required these days to implement direct uplod of quatized buffers.

**Comment by kring** [comment]
Thanks for the note @RaiaN! Just to be sure I'm understanding... you're saying you needed to modify the UE source code itself?

**Comment by RaiaN** [comment]
Hey @kring. Yeah, what I mean is that in case you wanted to partially update vertex/index buffer X in 4.2X, which is created by the engine itself (at PostLoad time), you would need to modify Lock_ functions as Epic didn't implement support for partial updates of "static" buffers. 

However, I assume you might be using BUF_Dynamic instead of BUF_Static so there should be no issues with partial updates in this case.

**Comment by csciguy8** [comment]
For reference, here's some of the latest related profiles in Unreal Insights...
![image](https://github.com/CesiumGS/cesium-unreal/assets/130494071/f1674f85-12b5-4689-b0d9-8dbad354f10e)
from [this PR](https://github.com/CesiumGS/cesium-unreal/pull/1356).

Setting up the mesh and material look to be around half the time spent in `::LoadPrimitive`, so if we can benefit from more asynchronous operation, fantastic.

Curiously, while we are using `RHIAsyncCreateTexture2D` for textures, it looks like we are using it in a function that waits for it to complete anyway? (kind of defeating the point)
https://github.com/CesiumGS/cesium-unreal/blob/95e3b227fc4138f9f3307c40c74493f6a4bbb218/Source/CesiumRuntime/Private/CesiumTextureUtility.cpp#L34

**Comment by kring** [comment]
> Curiously, while we are using RHIAsyncCreateTexture2D for textures, it looks like we are using it in a function that waits for it to complete anyway? (kind of defeating the point)

Waiting in a worker thread is much better than waiting in the game thread.

But in old versions of UE, the wait was "hidden", so it's possible we didn't even realize it was happening. It'd be interesting to explore whether we can eliminate it. But it doesn't defeat the point, in any case.

**Comment by csciguy8** [comment]
> Waiting in a worker thread is much better than waiting in the game thread.... But it doesn't defeat the point, in any case.

Ah, I think I see. 

So using `RHIAsyncCreateTexture2D` allows for a completion event later (which we don't currently take advantage of), but the other benefit is it can be called from a worker thread, where `RHICreateTexture2D` cannot?

Apologies for the naive questions, the UE docs on these calls are [a bit sparse](https://docs.unrealengine.com/4.26/en-US/API/Runtime/RHI/RHICreateTexture2D/), and [equally sparse](https://docs.unrealengine.com/4.26/en-US/API/Runtime/RHI/RHIAsyncCreateTexture2D/).

**Comment by kring** [comment]
> but the other benefit is it can be called from a worker thread, where RHICreateTexture2D cannot?

Right. It is my understanding that all of the `RHI*` functions, unless explicitly documented otherwise, can only be called from the render thread.
